dm/master/etcd.go

```diff
@@ -59,3 +68,111 @@ func startEtcd(masterCfg *Config,
 	}
 	return e, nil
 }
+
+// prepareJoinEtcd prepares config needed to join an existing cluster.
+// learn from https://github.com/pingcap/pd/blob/37efcb05f397f26c70cda8dd44acaa3061c92159/server/join/join.go#L44.
+func prepareJoinEtcd(cfg *Config) error {
+	// no need to join
+	if cfg.Join == "" {
+		return nil
+	}
+
+	// try to join self, invalid
+	if cfg.Join == cfg.AdvertisePeerUrls {
+		return terror.ErrMasterJoinEmbedEtcdFail.Generate(fmt.Sprintf("join self %s is forbidden", cfg.Join))
+	}
+
+	// join with persistent data
+	joinFP := filepath.Join(cfg.DataDir, "join")
+	if _, err := os.Stat(joinFP); !os.IsNotExist(err) {
+		s, err := ioutil.ReadFile(joinFP)
+		if err != nil {
+			return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, "read persistent join data")
+		}
+		cfg.InitialCluster = strings.TrimSpace(string(s))
+		cfg.InitialClusterState = embed.ClusterStateFlagExisting
+		return nil
+	}
+
+	// restart with previous data, no `InitialCluster` need to set
+	if isDataExist(filepath.Join(cfg.DataDir, "member")) {
+		cfg.InitialCluster = ""
+		cfg.InitialClusterState = embed.ClusterStateFlagExisting
+		return nil
+	}
+
+	// if without previous data, we need a client to contact with the existing cluster.
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:   strings.Split(cfg.Join, ","),
+		DialTimeout: etcdutil.DefaultDialTimeout,
+	})
+	if err != nil {
+		return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, fmt.Sprintf("create etcd client for %s", cfg.Join))
+	}
+	defer client.Close()
+
+	// `member list`
+	listResp, err := etcdutil.ListMembers(client)
+	if err != nil {
+		return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, fmt.Sprintf("list member for %s", cfg.Join))
+	}
+
+	// check members
+	for _, m := range listResp.Members {
+		if m.Name == "" {
+			return terror.ErrMasterJoinEmbedEtcdFail.Generate("there is a member that has not joined successfully")
+		}
+		if m.Name == cfg.Name {
+			// a failed DM-master re-joins the previous cluster.
+			return terror.ErrMasterJoinEmbedEtcdFail.Generate(fmt.Sprintf("missing data or joining a duplicate member %s", m.Name))
+		}
+	}
+
+	// `member add`, a new/deleted DM-master joins to an existing cluster.
+	addResp, err := etcdutil.AddMember(client, strings.Split(cfg.AdvertisePeerUrls, ","))
+	if err != nil {
+		return terror.ErrMasterJoinEmbedEtcdFail.Delegate(err, fmt.Sprintf("add member %s", cfg.AdvertisePeerUrls))
+	}
+
+	// generate `--initial-cluster`
+	ms := make([]string, 0, len(addResp.Members))
+	for _, m := range addResp.Members {
+		name := m.Name
+		if m.ID == addResp.Member.ID {
+			name = cfg.Name
```

Comment:
would `m.Name` be empty?

Reply:
It can be empty, I add some explanation in https://github.com/pingcap/dm/pull/350/commits/dcda7b23a1dad59bbdc8dd3d69e12d9cb89cc623.

Reply:
if name is empty, will it generate err at L142?